app/src/main/java/org/dhis2/usescases/jira/JiraViewModel.kt

```diff
@@ -1,208 +1,138 @@
-        return isSessionOpen
+    fun onDescriptionChanged(s: CharSequence, start: Int, before: Int, count: Int) {
+        description = s.toString()
+        checkFormCompletion()
     }
 
-    override fun sendIssue() {
-        val issueRequest = IssueRequest(summary.value, description.value)
-        val basic = String.format("Basic %s", session.value)
-        val requestBody =
-            RequestBody.create(MediaType.parse("application/json"), Gson().toJson(issueRequest))
-        issueService.createIssue(basic, requestBody).enqueue(getIssueCallback())
+    fun onJiraUserChanged(s: CharSequence, start: Int, before: Int, count: Int) {
+        userName = s.toString()
     }
 
-    private fun getIssueCallback(): Callback<ResponseBody>? {
-        return object : Callback<ResponseBody> {
-            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
-                if (response.isSuccessful) {
-                    issueMessage.value = "Issue Sent"
-                    getJiraIssues()
-                } else {
-                    issueMessage.value = "Error sending issue"
-                }
-            }
+    fun onJiraPassChanged(s: CharSequence, start: Int, before: Int, count: Int) {
+        pass = s.toString()
+    }
 
-            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
-                issueMessage.value = "Error sending issue"
-            }
-        }
+    fun onCheckedChanged(isChecked: Boolean) {
+        rememberCredentials = isChecked
     }
 
-    override fun onSummaryChanged(s: CharSequence, start: Int, before: Int, count: Int) {
-        summary.value = s.toString()
-        checkFormCompletion()
+    fun onJiraIssueClick(jiraKey: String) {
+        clickedIssueData.value = ClickedIssueData(
+            jiraKey.toJiraIssueUri(),
+            jiraRepository.getSession() ?: ""
+        )
     }
 
-    override fun onDescriptionChanged(s: CharSequence, start: Int, before: Int, count: Int) {
-        description.value = s.toString()
-        checkFormCompletion()
+    private fun saveCredentialsIfNeeded() {
+        if (rememberCredentials) {
+            jiraRepository.saveCredentials()
+        }
     }
 
-    override fun onJiraUserChanged(s: CharSequence, start: Int, before: Int, count: Int) {
-        userName.value = s.toString()
+    @VisibleForTesting
+    fun handleSendResponse(responseIsSuccessful: Boolean) {
+        if (responseIsSuccessful) {
+            issueMessage.value = resources.jiraIssueSentMessage()
+            getJiraTickets()
+        } else {
+            issueMessage.value = resources.jiraIssueSentErrorMessage()
+        }
     }
 
-    override fun onJiraPassChanged(s: CharSequence, start: Int, before: Int, count: Int) {
-        pass.value = s.toString()
+    @VisibleForTesting
+    fun handleThrowableResponse(throwable: Throwable) {
+        issueMessage.value = throwable.localizedMessage
```

Comment:
I don't think you need to use VisibleforTesting. With Mockito you can mocks callbacks and choose either success or error, then you can make the assertions. Let me know if you need help with this !

Reply:
true